src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -766,6 +766,17 @@ CloseRemoteTransaction(struct MultiConnection *connection)
 		/* XXX: Should we error out for a critical transaction? */
 
 		dlist_delete(&connection->transactionNode);
+
+		/*
+		 * If the transaction was completed, we have now cleaned it up, so we
+		 * can reset the state to REMOTE_TRANS_NOT_STARTED. This allows us to
+		 * start a new transaction without running into errors.
+		 */
+		if (transaction->transactionState == REMOTE_TRANS_ABORTED ||
+			transaction->transactionState == REMOTE_TRANS_COMMITTED)
+		{
+			transaction->transactionState = REMOTE_TRANS_NOT_STARTED;
+		}
```

Comment:
let's not touch CloseRemoteTransaction for now and use ResetConnection instead

Reply:
Removing this change and calling ResetConnection causes this assert to be hit in StartRemoteTransactionBegin
`	Assert(transaction->transactionState == REMOTE_TRANS_NOT_STARTED);
`

Reply:
Hmm, the situation is a little tricky.

We have a single `superuser` connection, and we sometimes use it  with `SendCommandListToWorkerOutsideTransactionWithConnection`, and sometimes inside a transaction `BlockWritesToShardList->LockShardListMetadataOnWorkers->SendCommandToWorkersWithMetadata->SendCommandToWorkersParamsInternal->RemoteTransactionsBeginIfNecessary`


Now, I'm trying to understand if this is a limitation of the connection management code (e.g., it gives a connection that it should not be giving) or misuse of the superuser connection.

Also, just sharing a commit where should get close to here to avoid additional connections: https://github.com/citusdata/citus/compare/reuse-connections-for-logical-ref-fkeys...reuse-connections-for-logical-ref-fkeys_ond?expand=1

Reply:
Wait, ignore above, we already have ClaimConnectionExclusively on the superuser connection. I think I broke it with adding ResetConnection, which unclaims the connection

Reply:
Ok, we need roughly something like this: https://github.com/citusdata/citus/compare/reuse-connections-for-logical-ref-fkeys...reuse-connections-for-logical-ref-fkeys_ond?expand=1#diff-dae583fd6d707aa7fd8ec0849dc8462a0766446152f84cfb7a56698edb1fae67R673


Given that we already tagged 11.1, this will go into patch release. Still, a pretty important fix IMO. Lets think a little more on Monday and make sure we do not miss anything.